beacon-chain/operations/voluntaryexits/pool.go

```diff
@@ -0,0 +1,167 @@
+	p.lock.RLock()
+	length := int(math.Min(float64(params.BeaconConfig().MaxVoluntaryExits), float64(p.pending.Len())))
+	result := make([]*ethpb.SignedVoluntaryExit, 0, length)
+	node := p.pending.Last()
+	for node != nil && len(result) < length {
+		exit, err := node.Value()
+		if err != nil {
+			p.lock.RUnlock()
+			return nil, err
+		}
+		if exit.Exit.Epoch > slots.ToEpoch(slot) {
+			node, err = node.Prev()
+			if err != nil {
+				p.lock.RUnlock()
+				return nil, err
+			}
+			continue
+		}
+		validator, err := state.ValidatorAtIndexReadOnly(exit.Exit.ValidatorIndex)
+		if err != nil {
+			p.handleInvalidExit(err, exit)
+		}
+		if err = blocks.VerifyExitConditions(validator, state.Slot(), exit.Exit); err != nil {
+			p.handleInvalidExit(err, exit)
+		}
+		result = append(result, exit)
+		node, err = node.Prev()
+		if err != nil {
+			p.lock.RUnlock()
+			return nil, err
+		}
+	}
+	p.lock.RUnlock()
+	if len(result) == 0 {
+		return result, nil
+	}
+	// We now verify the signatures in batches
+	cSet, err := blocks.ExitSignatureBatch(state, result)
+	if err != nil {
+		logrus.WithError(err).Warning("could not get exit signatures")
+	} else {
+		ok, err := cSet.Verify()
+		if err != nil {
+			logrus.WithError(err).Warning("could not batch verify exit signatures")
+		} else if ok {
+			return result, nil
+		}
+	}
+	// Batch signature failed, check signatures individually
+	verified := make([]*ethpb.SignedVoluntaryExit, 0, length)
+	for i, sig := range cSet.Signatures {
+		signature, err := blst.SignatureFromBytes(sig)
+		if err != nil {
+			logrus.WithError(err).Warning("could not get signature from bytes")
+			continue
+		}
+		if !signature.Verify(cSet.PublicKeys[i], cSet.Messages[i][:]) {
+			logrus.Warning("removing exit with invalid signature from pool")
+			p.MarkIncluded(result[i])
+		} else {
+			verified = append(verified, result[i])
+		}
+	}
+	return verified, nil
+}
+
+// InsertVoluntaryExit into the pool.
+func (p *Pool) InsertVoluntaryExit(exit *ethpb.SignedVoluntaryExit) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	_, exists := p.m[exit.Exit.ValidatorIndex]
+	if exists {
+		return
+	}
+
+	p.pending.Append(doublylinkedlist.NewNode(exit))
+	p.m[exit.Exit.ValidatorIndex] = p.pending.Last()
+}
+
+// MarkIncluded is used when an exit has been included in a beacon block. Every block seen by this
+// node should call this method to include the exit. This will remove the exit from the pool.
+func (p *Pool) MarkIncluded(exit *ethpb.SignedVoluntaryExit) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	node := p.m[exit.Exit.ValidatorIndex]
+	if node == nil {
+		return
+	}
+
+	delete(p.m, exit.Exit.ValidatorIndex)
+	p.pending.Remove(node)
+}
+
+func (p *Pool) handleInvalidExit(err error, exit *ethpb.SignedVoluntaryExit) {
+	logrus.WithError(err).Warning("removing invalid exit from pool")
+	// MarkIncluded removes the invalid exit from the pool
+	p.lock.RUnlock()
```

Comment:
I don't like the logic here: this function assumes that the caller has a lock on the pool and releases it and then calls `MarkIncluded` that takes back a lock. This is a recipe for disaster in any future refactor.